bsddialog.1
.\"
.\" Copyright (c) 2021-2023 Alfonso Sabato Siciliano
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd July 19, 2023
.Dt BSDDIALOG 1
.Os
.Sh NAME
.Nm bsddialog
.Nd TUI dialogs
.Sh SYNOPSIS
.Nm
.Fl Fl help | Fl Fl version
.Nm
.Op Fl Fl option
.Fl Fl dialog
.Ar text
.Ar rows
.Ar cols
.Op Ar dialog-argument
.Op Fl Fl option
.Nm
\&...
.Fl Fl dialog1
.Ar ...
.Oo Fl Fl and-dialog
.Fl Fl dialog2
.Ar ...
.Oc ...
.Sh DESCRIPTION
The
.Nm bsddialog
utility builds Text User Interface dialogs and widgets: to display messages,
to get input and to inform about a computation status.
.Pp
The options
.Fl Fl help
and
.Fl Fl version
print the list of options and the version, respectively, then exit.
.Fl Fl and-dialog
builds another dialog unless the previous one returns Error, ESC or Cancel.
.Pp
.Ar text
is a message printed inside the dialog.
.Ar rows
and
.Ar cols
are the height and width, 0 for autosize and -1 for fullscreen.
.Pp
The possible input got from the user interface is printed to standard error.
.Ss Options
The following options can change the default behavior of the utility and are
common to some dialog.
.Bl -tag -width Ds
.It Fl Fl alternate-screen
If available set alternate screen mode, see
.Xr terminfo 5 .
.It Fl Fl ascii-lines
Ascii characters to draw lines.
.It Fl Fl backtitle Ar backtitle
Title on the top side of the screen.
.It Fl Fl begin-x Ar x
Dialog horizontal position, 0 is the left screen side, -1 center.
.It Fl Fl begin-y Ar y
Dialog vertical position, 0 is the top screen side, -1 center.
.It Fl Fl bikeshed
Random settings.
Colors.
Delimiter and margins around the title.
Buttons always active or TAB to switch focus with input components, see
.Fl Fl switch-buttons .
Zero padding with time or date output.
Button delimiter.
Date format for
.Fl Fl datebox ,
see
.Fl Fl datebox-format .
.It Fl Fl cancel-exit-code Ar retval
Set an exit code value for the
.Dq Cancel
button.
.It Fl Fl cancel-label Ar label
Label for the
.Dq Cancel
button.
.It Fl Fl clear-dialog
Hide the dialog at exit.
.It Fl Fl clear-screen
Clear the screen, after the dialog exit if a dialog is built.
.It Fl Fl columns-per-row Ar columns
Try to set the number of columns for a row of text with autosizing; default
.Dv 10 .
.It Fl Fl cr-wrap
Keep new line in
.Ar text
also if it constains a
.Dq \en ,
see
.Fl Fl text-unchanged .
.It Fl Fl datebox-format Ar format
String to customize
.Fl Fl datebox
interface, possible values:
.Dq d/m/y ,
.Dq m/y/d ,
.Dq y/m/d .
.It Fl Fl date-format Ar format
String accepted by
.Xr strftime 3
to customize the output of
.Fl Fl datebox
and
.Fl Fl calendar
.It Fl Fl default-button Ar label
Focus on the button with
.Ar label
on startup.
.It Fl Fl default-item Ar name
Focus on the item with
.Ar name ,
for Checklist, Menu, Radiolist and Treeview.
.It Fl Fl default-no
Focus on
.Dq Cancel
or
.Dq \&No
button on startup.
.It Fl Fl disable-esc
Disable ESC key to quit.
.It Fl Fl error-exit-code Ar retval
Set an exit code value for the
.Nm
errors.
.It Fl Fl esc-exit-code Ar retval
Set an exit code value for the
.Dv ESC
key.
.It Fl Fl extra-button
Add a button with
.Dq Extra
label.
.It Fl Fl extra-exit-code Ar retval
Set an exit code value for the
.Dq Extra
button.
.It Fl Fl extra-label Ar label
Set
.Ar label
for the
.Dq Extra
button.
.It Fl Fl left1-button Ar label
Add a button with
.Ar label .
.It Fl Fl left1-exit-code Ar retval
Set an exit code for
.Fl Fl left1-button .
.It Fl Fl left2-button Ar label
Add a button with
.Ar label .
.It Fl Fl left2-exit-code Ar retval
Set an exit code for
.Fl Fl left2-button .
.It Fl Fl left3-button Ar label
Add a button with
.Ar label .
.It Fl Fl left3-exit-code Ar retval
Set an exit code for
.Fl Fl left3-button .
.It Fl Fl help-button
Add a button with
.Dq Help
label.
.It Fl Fl help-exit-code Ar retval
Set an exit code value for the
.Dq Help
button.
.It Fl Fl help-label Ar label
Set
.Ar label
for
.Dq Help
button.
.It Fl Fl help-print-name
Print the name of the focused item if the Help button is pressed also
with
.Fl Fl item-bottom-desc .
.It Fl Fl help-status
Print also the selected items if the
.Dq Help
button is pressed.
.It Fl Fl hfile Ar filename
Open
.Ar filename
in a Textbox if F1 key is pressed,
.It Fl Fl hline Ar string
Dialog subtitle.
.It Fl Fl hmsg Ar string
Open a Msgbox with
.Ar string
if the F1 key is pressed.
.It Fl Fl ignore
Do not exit with unknown options.
.It Fl Fl insecure
Print
.Sq *
to hide passwords while typing; whitespace otherwise.
.It Fl Fl item-bottom-desc
Set a help string for each item of a Checklist, Form, Menu, Mixedform,
Passwordform, Radiolist and Treeview to display at the bottom screen side.
.It Fl Fl item-depth
Specify a margin for items, available for Checklist, Menu and Radiolist.
.It Fl Fl item-prefix
Set a string to prefix each item of a Checklist, Menu, Radiolist or Treeview.
.It Fl Fl load-theme Ar file
Load theme from
.Ar file .
.It Fl Fl max-input Ar size
Maximum length of the input for
.Fl Fl inputbox
ans
.Fl Fl passwordbox ,
default 2048.
.It Fl Fl no-cancel
Do not show
.Dq Cancel
button.
.It Fl Fl no-descriptions
Do not display items desciption, for Checklist, Menu, Radiolist or Treeview;
mutually exclusive with
.Fl Fl no-names .
.It Fl Fl no-lines
Do not draw borders and lines.
.It Fl Fl no-names
Do not display items name, for Checklist, Menu and Radiolist; mutually exclusive
with
.Fl Fl no-descriptions .
.It Fl Fl no-ok
Do not draw
.Dq OK
button.
.It Fl Fl no-shadow
No not draw the shadow of the dialog.
.It Fl Fl ok-label Ar label
Set
.Ar label
for
.Dq OK
button.
.It Fl Fl ok-exit-code Ar retval
Set an exit code value for the
.Dq Ok
button.
.It Fl Fl normal-screen
If available set normal screen mode, see
.Xr terminfo 5 .
.It Fl Fl output-fd Ar fd
Print input from user interface to the specified file descriptor.
.It Fl Fl output-separator Ar sep
Set a sepator for the items in output, default whitespace.
.It Fl Fl print-maxsize
Screen size.
This option can be used without a dialog.
.It Fl Fl print-size
Print Dialog height and widget at exit.
.It Fl Fl print-version
Print version.
This option can be used without a dialog.
.It Fl Fl quoted
Quote items in output, default only when necessary.
.It Fl Fl right1-button Ar label
Add a button with
.Ar label .
.It Fl Fl right1-exit-code Ar retval
Set an exit code for
.Fl Fl right1-button .
.It Fl Fl right2-button Ar label
Add a button with
.Ar label .
.It Fl Fl right2-exit-code Ar retval
Set an exit code for
.Fl Fl right2-button .
.It Fl Fl right3-button Ar label
Add a button with
.Ar label .
.It Fl Fl right3-exit-code Ar retval
Set an exit code for
.Fl Fl right3-button .
.It Fl Fl save-theme Ar file
Save the current theme.
This option can be used without a dialog.
.It Fl Fl separate-output
Separate selected items with a new line and avoid to quote.
.It Fl Fl shadow
Show a pseudo shadow for the dialog, enabled by default.
.It Fl Fl single-quoted
Use single quote for items in output.
.It Fl Fl sleep Ar secs
Wait
.Ar secs
seconds to close the dialog.
.It Fl Fl stderr
Print input from user interface to standand error, default.
.It Fl Fl stdout
Print input from user interface to standard output.
.It Fl Fl switch-buttons
enables focus switching between buttons and input components pressing TAB,
otherwise buttons are always active and ENTER key closes the dialog.
Suitable for:
.Fl Fl form ,
.Fl Fl inputbox ,
.Fl Fl mixedform ,
.Fl Fl passwordbox ,
.Fl Fl passwordform ,
.Fl Fl timebox ,
.Fl Fl calendar
and
.Fl Fl datebox .
.It Fl Fl tab-escape
Replace
.Dq \et
with a tab in
.Ar text .
.It Fl Fl tab-len Ar spaces
Number of spaces to print a TAB in
.Ar text .
.It Fl Fl text-escape
Enable escapes in
.Ar text :
.Bl -column -compact
.It Dq \eZ0
black.
.It Dq \eZ1
red.
.It Dq \eZ2
green.
.It Dq \eZ3
yellow.
.It Dq \eZ4
blue.
.It Dq \eZ5
magenta.
.It Dq \eZ6
cyan.
.It Dq \eZ7
white.
.It Dq \eZb
bold.
.It Dq \eZB
disable bold.
.It Dq \eZd
Half bright.
.It Dq \eZD
disable half bright.
.It Dq \eZk
Blink.
.It Dq \eZK
disable blinking.
.It Dq \eZr
reverse foreground and background.
.It Dq \eZR
disable reverse.
.It Dq \eZs
Highlight.
.It Dq \eZS
disable highlighting.
.It Dq \eZu
underline.
.It Dq \eZU
disable underline.
.It Dq \eZn
disable each customization.
.El
.It Fl Fl text-unchanged
By default the
.Ar text
is changed before to be printed.
If it contains at least a
.Dq \en
each new line and TAB is converted to a space, subsequent spaces are merged.
Otherwise new line characters are preserved and a TAB becomes a space.
This option disable the
.Ar text
modification.
.It Fl Fl theme Ar theme
Set a graphical style: 3d, blackwhite or flat.
.It Fl Fl time-format Ar format
String accepted by
.Xr strftime 3
to customize the output of
.Fl Fl timebox .
.It Fl Fl timeout-exit-code Ar retval
Set an exit code value for the
.Fl Fl pause
timeout.
.It Fl Fl title Ar title
Dialog title.
.El
.Ss Dialogs
The following dialogs are available:
.Bl -tag -width Ds
.It Fl Fl calendar Ar text Ar rows Ar cols Op Ar day Ar month Ar year
Dialog to select a date.
.It Fl Fl checklist Ar text Ar rows Ar cols Ar menurows Oo Ar name Ar desc \
Ar status Oc ...
Checklist to select some item from a list via the SPACE key.
An item has a
.Ar name ,
.Ar desc
and a default
.Ar status
specified by
.Dq on
or
.Dq off .
The names of the selected items are printed to standard error.
.Ar menurows
is the graphical height of the list, 0 for autosize.
.It Fl Fl datebox Ar text Ar rows Ar cols Op Ar day Ar month Ar year
Dialog to select a date.
.It Fl Fl form Ar text Ar rows Ar cols Ar formrows Oo Ar label Ar ylabel \
Ar xlabel Ar init Ar yfield Ar xfield Ar fieldlen Ar maxletters Oc ...
Dialog to get a list of strings via forms.
A form has a
.Ar label
at the position
.Ar ylabel
and
.Ar xlabel ,
a field to get the input at the position
.Ar yfield
and
.Ar xfield
with graphical length
.Ar fieldlen ,
.Ar maxletters
is the maximum input length.
The field can be customized, if
.Ar fieldlen
is negative the field is read only and its absolute value is the field length.
If
.Ar maxletters
is 0 it is the absolute value of
.Ar fieldlen .
.Ar init
is a default value.
.Ar formrows
is the graphical height of the list,
.Dv 0
for autosize.
.It Fl Fl gauge Ar text Ar rows Ar cols Op Ar percentage
Dialog with a bar to shows
.Ar percentage ,
then it waits to read
.Dq XXX
from the standard input, then the first string replaces percentage and the
following strings replace
.Ar text
until the next
.Dq XXX ,
the loop ends reading
.Dv EOF .
.It Fl Fl infobox Ar text Ar rows Ar cols
Dialog without buttons to display a message and to exit immediately.
.It Fl Fl inputbox Ar text Ar rows Ar cols Op Ar init
Dialog to get a string in input,
.Ar init
is the default value.
.It Fl Fl menu Ar text Ar rows Ar cols Ar menurows Oo Ar name desc Oc ...
Builds a menu to select an item from a list, Space key is equivalent to Enter.
An item has a
.Ar name
and a
.Ar desc .
The name of the selected item is printed to standard error.
.Ar menurows
is the graphical height of the list, 0 for autosize.
.It Fl Fl mixedform Ar text Ar rows Ar cols Ar formrows Oo Ar label Ar ylabel \
Ar xlabel Ar init Ar yfield Ar xfield Ar fieldlen Ar maxletters Ar flag Oc ...
Dialog to get a list of strings via forms.
A form has a
.Ar label
at the position
.Ar ylabel
and
.Ar xlabel ,
a field to get the input with graphical length
.Ar fieldlen
at the position
.Ar yfield
and
.Ar xfield ,
.Ar maxletters
is the maximum input length,
.Ar init
is a default value,
.Ar flag
can be 0 for normal field, 1 to hide the typed characters and 2 to set the
field read only.
.Ar formrows
is the graphical height of the list,
.Dv 0
for autosize.
.It Fl Fl mixedgauge Ar text Ar rows Ar cols Ar mainperc Oo Ar minilabel \
Ar miniperc Oc ...
Dialog to show a main bar to represent
.Ar mainperc
from 0 to 100 and some mini bar with a
.Ar minilabel
string and a
.Ar miniperc
with value from 0 and 100 or negative to print a descriptive string: -1
.Dq Succeeded ,
-2
.Dq Failed ,
-3
.Dq Passed ,
-4
.Dq Completed ,
-5
.Dq Checked ,
-6
.Dq Done ,
-7
.Dq Skipped ,
-8
.Dq \&In Progress ,
-9
to hide
.Fa miniperc ,
-10
.Dq N/A ,
-11
.Dq Pending ,
otherwise
.Dq UNKNOWN .
.It Fl Fl msgbox Ar text Ar rows Ar cols
Dialog to diplay a message without the
.Dq Cancel
button.
UP, DOWN, HOME, END, PAGEUP and PAGEDOWN keys are availble to navigate the text.
.It Fl Fl passwordbox Ar text Ar rows Ar cols Op Ar init
Dialog to get a password,
.Ar init
is the default value.

CHANGELOG
WIP Next Version 1.0

	Utility:
	* add: comments to -save-theme output file.
	* add: blink, halfbright, highlight, for --save-theme and --load-theme.
	* add: --datebox-format <d/m/y|m/d/y|y/m/d> to set --datebox UI.
	* add: --help-print-items (--help-status becames its alias).
	* add: --text-escape (--colors becames its alias).
	* add: escapes \Zd, \Zk, \Zs, \ZD, \ZK, \ZS, for --text-escape.
	* add: env NO_COLOR, to set blackwhite (high contrast) theme.
	* add: $HOME/.bsddialog.conf to load a startup file theme.
	* add: env BSDDIALOG_THEMEFILE to set a startup theme file.
	* add: --left1-button <label>.
	* add: --left2-button <label>.
	* add: --left3-button <label>.
	* add: --right1-button <label>.
	* add: --right2-button <label>.
	* add: --right3-button <label>.
	* add: dynamic exit codes.
	    - add: --error-exit-code.
	    - add: --ok-exit-code.
	    - add: --cancel-exit-code.
	    - add: --help-exit-code.
	    - add: --extra-exit-code.
	    - add: --timeout-exit-code.
	    - add: --esc-exit-code.
	    - add: --left1-exit-code.
	    - add: --left2-exit-code.
	    - add: --left3-exit-code.
	    - add: --right1-exit-code.
	    - add: --right2-exit-code.
	    - add: --right3-exit-code.
	    - add: env BSDDIALOG_ERROR.
	    - add: env BSDDIALOG_OK.
	    - add: env BSDDIALOG_CANCEL.
	    - add: env BSDDIALOG_HELP.
	    - add: env BSDDIALOG_EXTRA.
	    - add: env BSDDIALOG_TIMEOUT.
	    - add: env BSDDIALOG_ESC.
	    - add: env BSDDIALOG_LEFT1.
	    - add: env BSDDIALOG_LEFT2.
	    - add: env BSDDIALOG_LEFT3.
	    - add: env BSDDIALOG_RIGHT1.
	    - add: env BSDDIALOG_RIGHT2.
	    - add: env BSDDIALOG_RIGHT3.
	* add: undocumented envs for bsdconfig(8) compatibility.
	    - env BSDDIALOG_COMPATRC for use_shadow setting.
	    - env BSDDIALOG_ITEM_HELP to add/set exit code.
	* change: rename themes --theme 3d|blackwhite|flat.
	* change: --no-names and --no-descriptions mutually exclusive.
	* change: quote only checklist output items if necessary.
	* change: dialogs with user input print always values except with ERROR,
	    ESC, Cancel. Previously the situation was quite heterogeneous.
	* improve: DIAGNOSTIC messages adding fmt string errors.
	* improve: disable theme setting (opt and env) with no color terminals.
	* improve: menus on|off status: strcasecmp, diagnostic, real off check.
	* improve: --bikeshed with button delimiter and --date-format.
	* improve: --textbox accepts button options.
	* improve: Forms with Help button, print "HELP" (like menus),
	    accept --help-list-items and --help-print-name.
	* fix: --load-theme attributes.
	* fix: --clear-screen with --and-dialog.
	* delete: --theme bsddialog (partially implemented).
	* delete: --esc-return-cancel (replaced by new env and option).
	* delete: --generic-button1 (replaced by --right1-button).
	* delete: --generic-button2 (replaced by --right2-button).
	* refactor: modularize in more files (main, cli, builders, theme).

	Library:
	* add: bsddialog_inmode().
	* add: bsddialog_clear(y) for utility --clear-screen.
	* add: bsddialog_refresh() for utility terminal mode switch options.
	* add: conf.date.format="d/m/y"|"m/d/y"|"y/m/d" to customize
	    bsddialog_datebox() UI (boxes) with a date format.
	* add: 'const char *end' to bsddialog_gauge().
	* add: draw focus on the shortcut-key-selected button at exit.
	* add: escapes for conf.text.highlight \Zd, \Zk, \Zs, \ZD, \ZK, \ZS.
	* add: flags for theme features.
	    - BSDDIALOG_BLINK.
	    - BSDDIALOG_HALFBRIGHT.
	    - BSDDIALOG_HIGHLIGHT.
	* add: generic buttons.
	    - conf.button.left1_label, BSDDIALOG_LEFT1 return value.
	    - conf.button.left2_label, BSDDIALOG_LEFT2 return value.
	    - conf.button.left3_label, BSDDIALOG_LEFT3 return value.
	    - conf.button.right1.label, BSDDIALOG_RIGHT1 return value.
	    - conf.button.right2.label, BSDDIALOG_RIGHT2 return value.
	    - conf.button.right3.label, BSDDIALOG_RIGHT3 return value.
	* improve: check (when possible) API pointers.
	* improve: circolar buttons with left and right keys for msgbox, yesno,
	    menus, rangebox and pause.
	* improve: bsddialog_textbox() manages conf.buttons.
	* improve: bsddialog_datebox() a box change affects the others as well.
	* improve: bsddialog_geterror() with fmt strings.
	* change: API NULL strings handled like "", except gauge *sep and *end.
	* change: menus and form less restrictive with text, hide text with
	    little screens (same behavior as other dialogs).
	* change: mixedgauge BSDDIALOG_MG_BLANK does not draw minibar but prints
	    minilabel. The change allows mixedgauge to add sections. To restore
	    the previous behavior setting minilabel to "".
	* change: check/set bsddialog_gauge() perc max 100.
	* change: check/set bsddialog_mixedgauge() mainperc max 100.
	* change: conf.menu.no_name and conf.menu.no_desc mutually exclusive.
	* change: bsddialog_pause() sec -> *sec to know remaining time at exit.
	* change: add *focusitem to bsddialog_form() like menus.
	* change: "pointer" values are always set except when BSDDIALOG_ERROR
	    occurs. Examples *yy/*mm/*ss or rangebox *value.
	    - delete conf.menu.on_without_ok.
	    - delete conf.form.value_without_ok.
	* rename: conf.text.highlight -> conf.text.escape.
	* fix: bsddialog_pause() elevation bar after resize.
	* fix: bsddialog_textbox() key '0'.
	* fix: timebox.c checksize (boxes width).
	* fix: extend menurows after shrinking.
	* fix: menu pad and form pad "re-expansion" after shrink and enlarge.
	* fix: shadow top-left corner (trick wresize() before wmove()).
	* fix: increment bsddialog_total_progview for i386 and more general use.
	* delete: BSDDIALOG_THEME_BSDDIALOG (partially implemented).
	* delete: conf.button.generic1_label (for new conf.button.right1_label).
	* delete: BSDDIALOG_GENERIC1 return value (new BSDDIALOG_RIGHT1).
	* delete: conf.button.generic2_label (for new conf.button.right2_label).
	* delete: BSDDIALOG_GENERIC2 return value (new BSDDIALOG_RIGHT2).
	* delete: bsddialog_clearterminal().

utility/bsddialog.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2021-2023 Alfonso Sabato Siciliano
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <getopt.h>
#include <limits.h>
#include <locale.h>
#include <signal.h>
#include <stdarg.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <term.h>

#include <bsddialog.h>
#include <bsddialog_theme.h>

#include "util.h"

#define EXITCODE(retval) (exitcodes[retval + 1].value)
#define UNUSED_PAR(x) UNUSED_ ## x __attribute__((__unused__))

static void custom_text(struct options *opt, char *text, char *buf);

/* Exit codes */
struct exitcode {
	const char *name;
	int value;
};

static struct exitcode exitcodes[14] = {
	{ "BSDDIALOG_ERROR",    255 },
	{ "BSDDIALOG_OK",         0 },
	{ "BSDDIALOG_CANCEL",     1 },
	{ "BSDDIALOG_HELP",       2 },
	{ "BSDDIALOG_EXTRA",      3 },
	{ "BSDDIALOG_TIMEOUT",    4 },
	{ "BSDDIALOG_ESC",        5 },
	{ "BSDDIALOG_LEFT1",      6 },
	{ "BSDDIALOG_LEFT2",      7 },
	{ "BSDDIALOG_LEFT3",      8 },
	{ "BSDDIALOG_RIGHT1",     9 },
	{ "BSDDIALOG_RIGHT2",    10 },
	{ "BSDDIALOG_RIGHT3",    11 },
	{ "BSDDIALOG_ITEM_HELP",  2 } /* like HELP by default */
};

void set_exit_code(int lib_retval, int exitcode)
{
	exitcodes[lib_retval + 1].value = exitcode;
}

/* Error */
void exit_error(bool usage, const char *fmt, ...)
{
	va_list arg_ptr;

	if (bsddialog_inmode())
		bsddialog_end();
	printf("Error: ");
	va_start(arg_ptr, fmt);
	vprintf(fmt, arg_ptr);
	va_end(arg_ptr);
	printf(".\n\n");
	if (usage) {
		printf("See \'bsddialog --help\' or \'man 1 bsddialog\' ");
		printf("for more information.\n");
	}

	exit (EXITCODE(BSDDIALOG_ERROR));
}

void error_args(const char *dialog, int argc, char **argv)
{
	int i;

	if (bsddialog_inmode())
		bsddialog_end();
	printf("Error: %s unexpected argument%s:", dialog, argc > 1 ? "s" : "");
	for (i = 0; i < argc; i++)
		printf(" \"%s\"", argv[i]);
	printf(".\n\n");
	printf("See \'bsddialog --help\' or \'man 1 bsddialog\' ");
	printf("for more information.\n");

	exit (EXITCODE(BSDDIALOG_ERROR));
}

/* init */
static void sigint_handler(int UNUSED_PAR(sig))
{
	bsddialog_end();

	exit(EXITCODE(BSDDIALOG_ERROR));
}

static void start_bsddialog_mode(void)
{
	if (bsddialog_inmode())
		return;
	if (bsddialog_init() != BSDDIALOG_OK)
		exit_error(false, bsddialog_geterror());

	signal(SIGINT, sigint_handler);
}

static void getenv_exitcodes(void)
{
	int i;
	int value;
	char *envvalue;

	for (i = 0; i < 10; i++) {
		envvalue = getenv(exitcodes[i].name);
		if (envvalue == NULL || envvalue[0] == '\0')
			continue;
		value = (int)strtol(envvalue, NULL, 10);
		exitcodes[i].value = value;
		/* ITEM_HELP follows HELP without explicit setting */
		if(i == BSDDIALOG_HELP + 1)
			exitcodes[BSDDIALOG_ITEM_HELP + 1].value = value;
	}
}

/*
 * bsddialog utility: TUI widgets and dialogs.
 */
int main(int argc, char *argv[argc])
{
	bool startup;
	int i, rows, cols, retval, parsed, nargc, firstoptind;
	char *text, **nargv, *pn;
	struct bsddialog_conf conf;
	struct options opt;

	setlocale(LC_ALL, "");
	getenv_exitcodes();
	firstoptind = optind;
	pn = argv[0];
	retval = BSDDIALOG_OK;

	for (i = 0; i < argc; i++) {
		if (strcmp(argv[i], "--version") == 0) {
			printf("Version: %s\n", LIBBSDDIALOG_VERSION);
			return (BSDDIALOG_OK);
		}
		if (strcmp(argv[i], "--help") == 0) {
			usage();
			return (BSDDIALOG_OK);
		}
	}

	startup = true;
	while (true) {
		parsed = parseargs(argc, argv, &conf, &opt);
		nargc = argc - parsed;
		nargv = argv + parsed;
		argc = parsed - optind;
		argv += optind;

		if (opt.mandatory_dialog && opt.dialogbuilder == NULL)
			exit_error(true, "expected a --<dialog>");

		if (opt.dialogbuilder == NULL && argc > 0)
			error_args("(no --<dialog>)", argc, argv);

		/* --print-maxsize or --print-version */
		if (opt.mandatory_dialog == false && opt.clearscreen == false &&
		    opt.savethemefile == NULL && opt.dialogbuilder == NULL) {
			retval = BSDDIALOG_OK;
			break;
		}

		/* --<dialog>, --save-theme or clear-screen */
		if (opt.dialogbuilder != NULL) {
			if (argc < 3)
				exit_error(true,
				    "expected <text> <rows> <cols>");
			if ((text = strdup(argv[0])) == NULL)
				exit_error(false, "cannot allocate <text>");
			if (opt.dialogbuilder != textbox_builder)
				custom_text(&opt, argv[0], text);
			rows = (int)strtol(argv[1], NULL, 10);
			cols = (int)strtol(argv[2], NULL, 10);
			argc -= 3;
			argv += 3;
		}

		/* bsddialog terminal mode (first iteration) */
		start_bsddialog_mode();

		if (opt.screen_mode != NULL) {
			opt.screen_mode = tigetstr(opt.screen_mode);
			if (opt.screen_mode != NULL &&
			    opt.screen_mode != (char*)-1) {
				tputs(opt.screen_mode, 1, putchar);
				fflush(stdout);
				bsddialog_refresh();
			}
		}

		/* theme */
		if (startup)
			startuptheme();
		startup = false;
		if ((int)opt.theme >= 0)
			setdeftheme(opt.theme);
		if (opt.loadthemefile != NULL)
			loadtheme(opt.loadthemefile, false);
		if (opt.bikeshed)
			bikeshed(&conf);
		if (opt.savethemefile != NULL)
			savetheme(opt.savethemefile);

		/* backtitle and dialog */
		if (opt.dialogbuilder == NULL)
			break;
		if (opt.backtitle != NULL)
			if(bsddialog_backtitle(&conf, opt.backtitle))
				exit_error(false, bsddialog_geterror());
		retval = opt.dialogbuilder(&conf, text, rows, cols, argc, argv,
		    &opt);
		free(text);
		if (retval == BSDDIALOG_ERROR)
			exit_error(false, bsddialog_geterror());
		if (conf.get_height != NULL && conf.get_width != NULL)
			dprintf(opt.output_fd, "DialogSize: %d, %d\n",
			    *conf.get_height, *conf.get_width);
		if (opt.clearscreen)
			bsddialog_clear(0);
		opt.clearscreen = false;
		/* --and-dialog ends loop with Cancel or ESC */
		if (retval == BSDDIALOG_CANCEL || retval == BSDDIALOG_ESC)
			break;
		argc = nargc;
		argv = nargv;
		if (argc <= 0)
			break;
		/* prepare next parseargs() call */
		argc++;
		argv--;
		argv[0] = pn;
		optind = firstoptind;
	}

	if (bsddialog_inmode()) {
		/* --clear-screen can be a single option */
		if (opt.clearscreen)
			bsddialog_clear(0);
		bsddialog_end();
	}
	/* end bsddialog terminal mode */

	return (EXITCODE(retval));
}

void custom_text(struct options *opt, char *text, char *buf)
{
	bool trim, crwrap;
	int i, j;

	if (strstr(text, "\\n") == NULL) {
		/* "hasnl" mode */
		trim = true;
		crwrap = true;
	} else {
		trim = false;
		crwrap = opt->cr_wrap;
	}
	if (opt->text_unchanged) {
		trim = false;
		crwrap = true;
	}

	i = j = 0;
	while (text[i] != '\0') {
		switch (text[i]) {
		case '\\':
			buf[j] = '\\';
			switch (text[i+1]) {
			case 'n': /* implicitly in "hasnl" mode */
				buf[j] = '\n';
				i++;
				if (text[i+1] == '\n')
					i++;
				break;
			case 't':
				if (opt->tab_escape) {
					buf[j] = '\t';
				} else {
					j++;
					buf[j] = 't';
				}
				i++;
				break;
			}
			break;
		case '\n':
			buf[j] = crwrap ? '\n' : ' ';
			break;
		case '\t':
			buf[j] = opt->text_unchanged ? '\t' : ' ';
			break;
		default:
			buf[j] = text[i];
		}
		i++;
		if (!trim || buf[j] != ' ' || j == 0 || buf[j-1] != ' ')
			j++;
	}
	buf[j] = '\0';
}
